### PR TITLE
gsplat-rust-renderer: beat Brush throughput, 60 FPS at 1M splats (rerun 0.33), scale-validated accuracy

### PR DESCRIPTION
**Pull request by pablo-rerun**
> **Update — rerun 0.33 + Spark/aarch64 pixi infra.** Since this PR was opened, the viewer moved **off the local rerun-0.34-dev checkout to published rerun 0.33**. The 0.34 + kittest-mcp checkout was MCP-debug only and is **no longer a build dependency** — this **resolves the P1 "default `cargo build` only works on this machine"** limitation noted below (no more absolute-path `re_*` deps); the MCP debug workflow is documented in the package README. The branch also carries a Spark/aarch64 + pixi-infra payload:
>
> - **CUDA stack consolidated** into `[feature.cuda]` — torch + onnxruntime-gpu + TensorRT in one place; per-package TensorRT/torchvision/pytorch dupes and the per-package `LD_LIBRARY_PATH` hack removed.
> - **DGX Spark / linux-aarch64 support** — onnxruntime CUDA EP via the Jetson AI Lab `sbsa/cu130` wheel; obsolete wilor-nano ARM→CPU guards dropped.
> - **onnxruntime-gpu pinned `==1.24.0`** on both platforms (no x86/aarch64 skew); `requires-pixi >=0.70`.
> - **dpretrieval** `libopencv <5` (OpenCV 5 broke the DBoW2 build / any relock).
> - Removed the broken local-path `daggr` dep; `simplecv` `RerunTyroConfig.recording_id` is now CLI-configurable.
> - **gsplat docs trimmed** — removed ~4.4 MB of embedded-media HTML + the stale testing-plan; `architecture.md` verified against the code.
> - pixi **0.71 evaluated and deferred** (rich-platforms = a breaking, manifest-wide migration); workspace stays on **0.70.2**.
>
> _The original description below predates the 0.33 move — "rerun 0.34-alpha", the local checkout, and the P1 local-build limitation are historical and no longer apply._

---

## Summary

This branch takes the `gsplat-rust-renderer` package from "correct port" to "faster than the reference, validated at 1M-splat scale, running on rerun 0.34-alpha with MCP support."

### Performance vs Brush (RTX 5090)

| | small scenes (bench protocol) | T&T train, 1.07M splats (viewer, interactive orbit) |
|---|---|---|
| Brush | baseline | 24–33 ms (30–41 FPS, instrumented brush-app @ main) |
| **this branch** | **1.29–1.58× faster** (all 6 bench cells) | **16.6 ms (60 FPS, vsync-capped)** |

The viewer headline came from one root-cause find: rerun instantiates visualizer systems fresh every frame, so the per-instance cloud cache never survived a frame — every frame silently rebuilt and re-uploaded the full cloud (invisible on small scenes, 228 ms at 1M splats). The cache now lives in the store's `Memoizers` (`CloudCache`: store-event invalidation, memory-pressure purge, memory-panel reporting), with matching idle-eviction on the GPU side.

### Accuracy

- Bit-identical renders maintained through every perf phase (24-view gate + calibration + 310k-splat cloud).
- 1M-splat validation on Tanks & Temples *train* (official INRIA 30k PLY, standard every-8th eval protocol, COLMAP→NeRF camera conversion fed identically to both renderers): **ours-vs-GT ≥ 0.992× Brush-vs-GT on every view** (gate 0.98×), ours-vs-Brush mean 52.0 dB / SSIM 0.997, absolute 22.4 dB vs GT matches the published 3DGS number. Report §10.5.
- Re-log staleness fixed (content-hash signature + generation-driven GPU re-upload), with `tools/relog_check.py` as a regression gate.

### Viewer / ecosystem

- Builds against the local rerun checkout (0.34.0-alpha.1+dev) with the kittest-mcp egui patches — the live window is drivable/inspectable via the `re-mcp` MCP server (verified: camera orbit/zoom without blueprints, UI clicks by text, screenshots).
- Stock-archetype interop demonstrated: dataset cameras as `rr.Pinhole`/`rr.Transform3D` (simplecv pinholes; RUB and RDF conventions both pass through unmodified) + GT `rr.Image`s coexisting with the splat visualizer.
- New `tools/log_splats_with_cameras.py` (Tyro CLI, `RerunTyroConfig`, NeRF-synthetic + COLMAP autodetect) with two image-panel modes: `tabs` (4 named camera views per tab — individually inspectable; ~0.2 ms/frame per blueprint view) and `pages` (4-view timeline pager, 60 FPS at any camera count).
- Two measured viewer-perf findings documented for upstream: every blueprint view costs ~0.2 ms/frame whether or not its tab is visible, and 0.33-SDK `EncodedImage` triggers per-frame JPEG re-decode in the 0.34-alpha viewer (decode-cache miss under version skew).

### Code health

- −12% LOC alongside the perf work; two three-reviewer simplify passes and two code reviews applied (GPU cache eviction, build-outside-lock, lineage-filtered invalidation, shared probe gating).
- Env-gated diagnostics (`GSPLAT_FPS_PROBE=1`): frame-period EMA, intersection counts, rebuild tracing.
- Docs: `docs/implementation-notes.html` (decision log + embedded orbit video, MCP-control video, interop screenshots), `docs/brush-comparison.html` (full methodology + per-view metrics), `docs/architecture.md`.

## Known limitation (deliberate, needs follow-up)

The viewer feature's `re_*` dependencies are **absolute path deps into a local rerun checkout** (`/home/pablo/0Dev/work/rerun-projects/reality/rerun`) with mirrored branch-form `[patch.crates-io]` egui forks — required for the 0.34-alpha + kittest-mcp combination, but it means the default `cargo build` only works on this machine (flagged as P1 by Codex review). Options before wider use: vendor the checkout path behind an env var / `.cargo/config.toml`, or pin git deps once the MCP patches land upstream. The `--no-default-features` CLI build is unaffected.

## Validation

- `cargo clippy -D warnings` + 11 tests on both feature sets; ruff/pyrefly clean (pyrefly fix included for this package's `rerun` import resolution)
- Bit-identity + parity gates green at every commit that touched render code
- Live-viewer validation: relog check PASS, matched-pose screenshot SSIM 1.0000, 60 FPS probe-measured on the 1M-splat scene

### CODE REVIEW

File: packages/simplecv/simplecv/rerun_log_utils.py [hunk 1]
```diff
@@ -135,8 +134,10 @@ def get_safe_application_id() -> str:
...
     application_id: str = field(default_factory=get_safe_application_id)
     """Name of the application"""
-    recording_id: str | UUID | None = None
-    """Recording ID"""
+    recording_id: str | None = None
+    """Recording ID; pin one (any stable string) so separate processes/restarts
+    land in the same recording instead of forking new ones (tyro cannot parse a
+    UUID union member, which made this field non-configurable from the CLI)."""
     connect: bool = False
     """Wether to connect to an existing rerun instance or not"""
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/simplecv/simplecv/rerun_log_utils.py]: Fixed in 14eba19 (Wether -> Whether).

File: pixi.toml [hunk 2]
```diff
@@ -130,8 +144,47 @@ libnvjitlink = "*"
...
+
+# GPU runtime libs on LD_LIBRARY_PATH for every cuda env: $CONDA_PREFIX/lib first
+# (cuDNN/cuBLAS/cudart from conda — needed by ORT's CUDA & TRT providers), then the pip
+# TensorRT wheels' libnvinfer under site-packages/tensorrt_libs. Both py3.11 and py3.12
+# site-packages dirs are listed because cuda spans both python versions (e.g. dpvo and
+# mast3r-slam are 3.11); the loader silently skips whichever path doesn't exist. Setting
+# this key replaces pixi's default, so $CONDA_PREFIX/lib MUST be listed explicitly or
+# cuDNN won't resolve.
+[feature.cuda.activation.env]
+LD_LIBRARY_PATH = "$CONDA_PREFIX/lib:$CONDA_PREFIX/lib/python3.12/site-packages/tensorrt_libs:$CONDA_PREFIX/lib/python3.11/site-packages/tensorrt_libs:$LD_LIBRARY_PATH"
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply pixi.toml]: Fixed in 14eba19: the inherited path is now appended conditionally so an unset variable no longer leaves a trailing colon (which the loader treats as CWD). Verified TensorRT + ORT still import.

File: pixi.toml [hunk 3]
```diff
@@ -1478,15 +1502,30 @@ xorg-libxinerama = "*"
...
 lpips = ">=0.1.4,<1"
 plyfile = ">=1.1,<2"
-rerun-sdk = "==0.32.2"
+rerun-sdk = "==0.33.0"
 simplecv = { path = "packages/simplecv", editable = true }
 
 [feature.gsplat-rust-renderer.activation.env]
 PACKAGE_DIR = "packages/gsplat-rust-renderer"
+# Make pixi's pkg-config files (libudev.pc) visible to cargo builds.
+PKG_CONFIG_PATH = "${CONDA_PREFIX}/lib/pkgconfig"
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply pixi.toml:1528]: Intentional: pixi environments isolate from the host, so PKG_CONFIG_PATH is deliberately scoped to the env pkgconfig for reproducible cargo builds. Not changing.

File: pixi.toml [hunk 4]
```diff
@@ -94,13 +97,24 @@ prerelease-mode = "allow"
...
 
 [feature.cuda.system-requirements]
 cuda = "13.0"
+# CUDA-13 aarch64 index wheels ship as manylinux_2_3x: onnxruntime-gpu is _2_34,
+# the TensorRT aarch64 wheels (tensorrt-cu13-libs/bindings) are _2_35. Declare
+# glibc 2.35 so both resolve (the Spark runs 2.39). pixi system-requirements can't
+# be target-scoped, so this applies to both platforms — but it only *loosens*
+# package eligibility (every CUDA-13 host has glibc >= 2.35), so the existing
+# linux-64 solutions stay valid and are not re-solved.
+libc = { family = "glibc", version = "2.35" }
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply pixi.toml:117]: By design: the aarch64 TensorRT wheels are manylinux_2_35 and pixi system-requirements cannot be target-scoped, so glibc 2.35 applies to both platforms. All supported hosts are glibc >= 2.35 and the trade-off is documented in the adjacent comment. Not changing.

File: packages/gsplat-rust-renderer/tools/visualize_brush_training.py [hunk 5]
```diff
@@ -0,0 +1,840 @@
+"""Render real GPU splats inside brush's own training blueprint.
+
+brush-cli (main branch, --rerun-enabled) already logs a rich rerun blueprint —
+loss/lr/psnr/ssim/splats/refine/memory time-series, per-view (GT | Render) eval
+pairs, dataset cameras, the world frame — but its rerun SDK only knows coarse
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/tools/visualize_brush_training.py:5]: Valid per the tools/ thin-shim convention (Config + main belong in the package so beartype instruments them). Deferring the refactor of these 4 auxiliary demo/eval CLIs to a focused follow-up to keep this renderer PR scoped.

File: packages/gsplat-rust-renderer/tools/log_splats_with_cameras.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
+"""Log a trained splat PLY together with its dataset cameras and GT images.
+
+Demonstrates the custom ``GaussianSplats3D`` visualizer coexisting with stock
+rerun archetypes: every dataset camera becomes an ``rr.Pinhole`` frustum with
+its GT photo on the image plane, and the photos are also browsable in a paged
+2x2 grid driven by a ``page`` sequence timeline (scrub or play to flip
+through them — hidden tab views are NOT free in rerun, ~0.2 ms/frame each,
+so tabs of hundreds of views would tank the frame rate).
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/tools/log_splats_with_cameras.py:8]: Valid per the tools/ thin-shim convention (Config + main belong in the package so beartype instruments them). Deferring the refactor of these 4 auxiliary demo/eval CLIs to a focused follow-up to keep this renderer PR scoped.

File: packages/gsplat-rust-renderer/tools/calibration_scene.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+* four corner markers in the z=0 plane — focal length / FOV via inter-blob
+  pixel distances
+* an overlapping front/back pair on one camera ray (orange in front of purple)
+  — depth ordering / alpha compositing
+
+Usage:
+    python tools/calibration_scene.py generate --out-dir data/calibration
+    python tools/calibration_scene.py check --image render.png \
+        --scene-dir data/calibration --renderer gsplat-render
+"""
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/tools/calibration_scene.py:20]: Valid per the tools/ thin-shim convention (Config + main belong in the package so beartype instruments them). Deferring the refactor of these 4 auxiliary demo/eval CLIs to a focused follow-up to keep this renderer PR scoped.

File: packages/gsplat-rust-renderer/tools/relog_check.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
+"""Re-log regression check for the gsplat viewer (requires a GPU + built viewer binary).
+
+Logs a splat cloud, screenshots it, then RE-LOGS the same entity twice —
+(a) same splat count with different colors/positions, (b) different splat
+count — and asserts the rendered output follows the new data each time.
+
+Exercises the bug where the viewer's per-entity GPU caches (CloudSignature /
+batch_cache attribute buffers) served stale data after a re-log.
+"""
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/tools/relog_check.py:9]: Valid per the tools/ thin-shim convention (Config + main belong in the package so beartype instruments them). Deferring the refactor of these 4 auxiliary demo/eval CLIs to a focused follow-up to keep this renderer PR scoped.

File: packages/gsplat-rust-renderer/tools/run_brush_native_demo.sh [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+DATA_DIR="${1:?usage: run_brush_native_demo.sh DATA_DIR [TOTAL_ITERS] [EXPORT_DIR]}"
+TOTAL_ITERS="${2:-30000}"
+EXPORT_DIR="${3:-/tmp/brush-runs/$(basename "$DATA_DIR")}"
+EXPORT_EVERY="${EXPORT_EVERY:-200}"
+EVAL_EVERY="${EVAL_EVERY:-500}"
+# COLMAP captures (no transforms_val) need an eval split: brush holds out every
+# Nth view and the sidecar must match it so the eval-tab count lines up. 0 = off.
+EVAL_SPLIT_EVERY="${EVAL_SPLIT_EVERY:-0}"
+BRUSH_CLI="${BRUSH_CLI:-/home/pablo/0Dev/work/brush/target/release/brush-cli}"
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/tools/run_brush_native_demo.sh]: Fixed in 14eba19: defaults BRUSH_CLI to brush-cli (PATH) instead of an absolute path, so it works out of the box.

File: packages/gsplat-rust-renderer/Cargo.toml [hunk 10]
```diff
@@ -45,24 +47,30 @@ glam = "0.30"
...
-re_view_spatial = { version = "=0.31.1", optional = true }
-re_viewer = { version = "=0.31.1", optional = true }
-re_viewer_context = { version = "=0.31.1", optional = true }
-rerun = { version = "=0.31.1", features = ["native_viewer"], optional = true }
+# Published Rerun 0.33 crates from crates.io — reproducible from a clean
+# checkout on any machine.  The MCP viewport-control demo was developed
+# against a local 0.34.0-alpha rerun checkout + egui MCP-patch fork (path
+# deps + a [patch.crates-io] block); that setup is intentionally NOT shipped.
+# See README → "MCP viewport control (debug-only)" to reproduce it locally.
+egui_kittest = { version = "0.34.3", features = ["wgpu", "eframe"], optional = true }
```
[the remark this answers is not in the document]
Comment (pablo-rerun) [reply packages/gsplat-rust-renderer/Cargo.toml:60]: Addressed: the PR title/description were updated to reflect published rerun 0.33 (the 0.34-alpha checkout was MCP-debug-only, now documented in the README).
